### PR TITLE
fix(ivy): o2+ should work with multiple template instances

### PR DESCRIPTION
**Pull request by kara**
This PR fixes a bug where object/array literals with multiple expressions would not work properly with multiple template instances. Each template should have its own copy of the object/array literal, so the blueprint has been moved from TView into the next available slot in LView.data[].

cc @jelbourn @chuckjaz @vicb

### CODE REVIEW

File: packages/core/src/render3/instructions.ts [hunk 1]
```diff
@@ -1743,6 +1742,26 @@ function valueInData<T>(data: any[], index: number, value?: T): T {
...
+/** Wrapper for memory instruction to account for binding index (when outside creation mode block)
+ */
+function memoryWithBinding<T>(value?: T): T {
+  initBindings();
+  return memory<T>(bindingIndex++, value);
+}
+
+/** Copies an object or array */
+export function copyObject(obj: any): any {
+  return Array.isArray(obj) ? obj.slice() : {...obj};
```
Comment (mhevery) [inline packages/core/src/render3/instructions.ts]: can we use `Object.create` https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/create It should be more efficient that `{...obj}`

File: packages/core/src/render3/object_literal.ts [hunk 2]
```diff
@@ -42,43 +43,23 @@ function updateBinding4(
...
-  obj = getMutableBlueprint(objIndex, obj);
+export function objectLiteral1(obj: any, key: string | number, exp: any): any {
+  obj = getMutableBlueprint(obj);
   if (bind(exp) === NO_CHANGE) {
     return NO_CHANGE;
   } else {
     obj[key] = exp;
-    // Must copy to change identity when binding changes
+    // Must copy to change identity when binding changes for backwards compatibility
+    // Also supports deeply nested expressions
```
Comment (mhevery) [inline packages/core/src/render3/object_literal.ts]: What do you mean by `deeply`? To me it implies that we copy object deeply, which is not true.

Comment (mhevery) [reply packages/core/src/render3/object_literal.ts]: or do you mean `{ a: {b:exp} }` ? If so this comment is confusing. Can you make it more explicit and add test?

**Review by mhevery — CHANGES_REQUESTED** [review]

File: packages/core/src/render3/object_literal.ts (hunk 2 above)
Comment (kara) [reply packages/core/src/render3/object_literal.ts]: We have a unit test and a canonical test for nested expressions already. Happy to clarify the comment though. How about: ```Also supports nested expressions like {a: {b:exp}}```

File: packages/core/src/render3/instructions.ts (hunk 1 above)
Comment (kara) [reply packages/core/src/render3/instructions.ts]: What if there is a `hasOwnProperty` check in the component?

File: packages/core/src/render3/object_literal.ts (hunk 2 above)
Comment (vicb) [reply packages/core/src/render3/object_literal.ts]: @mhevery I like the discussion we had this morning about the semantic `[...]={}/[]` should have. Would be nice to put that down at some point.

File: packages/core/test/render3/compiler_canonical_spec.ts [hunk 3]
```diff
@@ -226,7 +226,7 @@ describe('compiler specification', () => {
                 r3.E(0, MyComp);
                 r3.e();
               }
-              r3.p(0, 'names', r3.o1(0, e0_literal, 1, ctx.customName));
```
Comment (vicb) [inline packages/core/test/render3/compiler_canonical_spec.ts]: nit: I would actually remove the tests here (`compiler_canonical_spec.ts`) until we fix the minification pb. So that @chuckjaz don't have to create and then change the compiler code.

Comment (kara) [reply packages/core/test/render3/compiler_canonical_spec.ts]: Eh, seems easier to just let Chuck know. Let's update the test when we have an update.

File: packages/core/src/render3/instructions.ts [hunk 4]
```diff
@@ -1743,6 +1742,26 @@ function valueInData<T>(data: any[], index: number, value?: T): T {
   return value !;
 }
 
+/**
+ * Gets a blueprint of an object or array if one has already been saved, or copies the
+ * object and saves it for the next change detection run if it hasn't.
+ */
+export function getMutableBlueprint(obj: any): any {
+  return creationMode ? memoryWithBinding(copyObject(obj)) : memoryWithBinding();
```
Comment (vicb) [inline packages/core/src/render3/instructions.ts]: add a note about `copyObject` -> needed to avoid cross instance pollution

File: packages/core/src/render3/instructions.ts [hunk 5]
```diff
@@ -1743,6 +1742,26 @@ function valueInData<T>(data: any[], index: number, value?: T): T {
...
+}
+
+/** Wrapper for memory instruction to account for binding index (when outside creation mode block)
+ */
+function memoryWithBinding<T>(value?: T): T {
+  initBindings();
+  return memory<T>(bindingIndex++, value);
```
Comment (vicb) [inline packages/core/src/render3/instructions.ts]: nit: do we need `<T>` here ?

File: packages/core/src/render3/instructions.ts [hunk 6]
```diff
@@ -1743,6 +1742,26 @@ function valueInData<T>(data: any[], index: number, value?: T): T {
...
 
+/**
+ * Gets a blueprint of an object or array if one has already been saved, or copies the
+ * object and saves it for the next change detection run if it hasn't.
+ */
+export function getMutableBlueprint(obj: any): any {
+  return creationMode ? memoryWithBinding(copyObject(obj)) : memoryWithBinding();
+}
+
+/** Wrapper for memory instruction to account for binding index (when outside creation mode block)
```
Comment (vicb) [inline packages/core/src/render3/instructions.ts]: No sure to get the comment here ?

File: packages/core/test/render3/object_literal_spec.ts [hunk 7]
```diff
@@ -30,7 +30,7 @@ describe('array literals', () => {
         E(0, MyComp);
         e();
       }
-      p(0, 'names', o1(0, e0_literal, 1, ctx.customName));
+      p(0, 'names', o1(e0_literal, 1, ctx.customName));
```
Comment (vicb) [inline packages/core/test/render3/object_literal_spec.ts]: Off topic: this is the kind of unit test that I think should be simplified.
Having:
1. To write a full template fn,
2. to call RenderToHtml

only to test calling `o1` is a lot of noise that IMO hurt readability hence maintainability

**Review by vicb — APPROVED** [review]
A few nits otherwise great change.

File: packages/core/src/render3/instructions.ts (hunk 6 above)
Comment (kara) [reply packages/core/src/render3/instructions.ts]: Will update with a longer explanation

**Review by mhevery — APPROVED** [review]

File: packages/core/test/render3/object_literal_spec.ts (hunk 7 above)
Comment (mhevery) [reply packages/core/test/render3/object_literal_spec.ts]: I am all for simpler tests. The complication here is that to test the instruction you have to correctly set up the instruction's state. Without it, the test is useless. Do you have ideas how to simplify and still know that correct state has been set up?

File: packages/core/src/render3/instructions.ts [hunk 8]
```diff
@@ -1743,6 +1742,11 @@ function valueInData<T>(data: any[], index: number, value?: T): T {
   return value !;
 }
 
+/** Gets the binding at the current bindingIndex and increments index. */
+export function getCurrentBinding(): any {
```
Comment (mhevery) [inline packages/core/src/render3/instructions.ts]:
```
export function peekBinding(): any {
  return data[bindingIndex];
}
```

File: packages/core/src/render3/object_literal.ts [hunk 9]
```diff
@@ -6,268 +6,206 @@
...
-}
-
-/** Updates four expressions in an object or array literal if they have changed. */
-function updateBinding4(
-    obj: any, key1: string | number, exp1: any, key2: string | number, exp2: any,
-    key3: string | number, exp3: any, key4: string | number, exp4: any): boolean {
-  let different = updateBinding2(obj, key1, exp1, key2, exp2);
-  return updateBinding2(obj, key3, exp3, key4, exp4) || different;
-}
+let different = false;
```
Comment (mhevery) [inline packages/core/src/render3/object_literal.ts]: I think there is a better way to do this. Revert this change.

File: packages/core/src/render3/object_literal.ts [hunk 10]
```diff
@@ -6,268 +6,206 @@
...
-  const tView = getTView();
-  const objectLiterals = tView.objectLiterals;
-  if (objectLiterals && index < objectLiterals.length) {
-    return objectLiterals[index];
-  } else {
-    ngDevMode && objectLiterals && assertEqual(index, objectLiterals.length, 'index');
-    return (objectLiterals || (tView.objectLiterals = []))[index] = copyObject(obj);
-  }
-}
+function getLatestValue(value: any): any {
```
Comment (mhevery) [inline packages/core/src/render3/object_literal.ts]: revert

File: packages/core/src/render3/object_literal.ts [hunk 11]
```diff
@@ -6,268 +6,206 @@
...
  * @param exp2
- * @returns A copy of the array or NO_CHANGE
+ * @returns A copy of the object/array or NO_CHANGE
  */
-export function objectLiteral2(
-    objIndex: number, obj: any, key1: string | number, exp1: any, key2: string | number,
-    exp2: any): any {
-  obj = getMutableBlueprint(objIndex, obj);
-  return updateBinding2(obj, key1, exp1, key2, exp2) ? copyObject(obj) : NO_CHANGE;
+export function objectLiteral2(factoryFn: (v1: any, v2: any) => any, exp1: any, exp2: any): any {
```
Comment (mhevery) [inline packages/core/src/render3/object_literal.ts]:
```
let different = false;
let peekValue;

peekValue = peekBinding();
if ((exp1 === NO_CHANGE || (exp1 = bind(exp1)) == NO_CHANGE) exp1 = peekValue else different = true;

peekValue = peekBinding();
if ((exp2 === NO_CHANGE || (exp2 = bind(exp2)) == NO_CHANGE) exp2 = peekValue else different = true;

return different ? factoryFn(exp1, exp2) : NO_CHANGE;
```

**Review by mhevery — CHANGES_REQUESTED** [review]
Also we need `objectLiteralV`

**Review by mhevery — APPROVED** [review]

File: packages/core/src/render3/object_literal.ts [hunk 12]
```diff
@@ -6,268 +6,285 @@
...
-    return NO_CHANGE;
-  } else {
-    obj[key] = exp;
-    // Must copy to change identity when binding changes
-    return copyObject(obj);
-  }
+export function objectLiteral1(factoryFn: (v: any) => any, exp: any): any {
+  let different = false;
+  if (exp === NO_CHANGE) exp = peekBinding();
+  if (bind(exp) !== NO_CHANGE) different = true;
```
Comment (vicb) [inline packages/core/src/render3/object_literal.ts]: `peekBinding()` might use the `bindingIndex` before it gets initialized no ?
Please add a test

**Review by vicb — CHANGES_REQUESTED** [review]
see inline comments

File: packages/core/src/render3/object_literal.ts (hunk 12 above)
Comment (vicb) [reply packages/core/src/render3/object_literal.ts]: IIUC the intent of this code isn't `exp = peekBinding();` harmful for the perf of `bind(exp)` ?

File: packages/core/test/render3/compiler_canonical_spec.ts [hunk 13]
```diff
@@ -226,7 +226,7 @@ describe('compiler specification', () => {
...
+              r3.p(0, 'names', r3.o1(e0_ff, ctx.customName));
```
Comment (vicb) [inline packages/core/test/render3/compiler_canonical_spec.ts]: This file is missing a quote vs unquoted spec.

File: packages/core/test/render3/object_literal_spec.ts [hunk 14]
```diff
@@ -162,42 +157,40 @@ describe('array literals', () => {
...
-          objectLiteral6(3, e6_literal, 2, c[2], 3, c[3], 4, c[4], 5, c[5], 6, c[6], 7, c[7]));
-      elementProperty(
-          8, 'names',
-          objectLiteral7(
-              4, e8_literal, 1, c[1], 2, c[2], 3, c[3], 4, c[4], 5, c[5], 6, c[6], 7, c[7]));
+      elementProperty(0, 'names', objectLiteral3(e0_ff, c[5], c[6], c[7]));
+      elementProperty(2, 'names', objectLiteral4(e2_ff, c[4], c[5], c[6], c[7]));
+      elementProperty(4, 'names', objectLiteral5(e4_ff, c[3], c[4], c[5], c[6], c[7]));
+      elementProperty(6, 'names', objectLiteral6(e6_ff, c[2], c[3], c[4], c[5], c[6], c[7]));
+      elementProperty(8, 'names', objectLiteral7(e8_ff, c[1], c[2], c[3], c[4], c[5], c[6], c[7]));
```
Comment (vicb) [inline packages/core/test/render3/object_literal_spec.ts]: having to do that for tests make me want to 😢 
(improving our tests is out of scope of this PR though)

File: packages/core/src/render3/instructions.ts [hunk 15]
```diff
@@ -1743,6 +1742,11 @@ function valueInData<T>(data: any[], index: number, value?: T): T {
   return value !;
 }
 
+/** Gets the binding at the current bindingIndex */
+export function peekBinding(): any {
+  return data[bindingIndex];
```
Comment (vicb) [inline packages/core/src/render3/instructions.ts]: `ngDevMode && assertNotSame(bindingIndex, null)`;

_not sure if it should be `null` or `undefined`_

File: packages/core/src/render3/object_literal.ts (hunk 12 above)
Comment (kara) [reply packages/core/src/render3/object_literal.ts]: The only time the `peekBinding` call is executed is if an `o()` instruction has already been run at a level above it, so it will never be uninitialized.

File: packages/core/test/render3/compiler_canonical_spec.ts (hunk 13 above)
Comment (kara) [reply packages/core/test/render3/compiler_canonical_spec.ts]: Not sure what you mean? Can you clarify?

File: packages/core/src/render3/instructions.ts (hunk 15 above)
Comment (kara) [reply packages/core/src/render3/instructions.ts]: The `bindingIndex` is always defined for a view (0 when uninitialized). I could add an assertion here for `bindingStartIndex` if you want to check that?

File: packages/core/src/render3/instructions.ts [hunk 16]
```diff
@@ -1743,6 +1742,12 @@ function valueInData<T>(data: any[], index: number, value?: T): T {
   return value !;
 }
 
+/** Gets the binding at the current bindingIndex */
+export function peekBinding(): any {
+  ngDevMode && assertNotEqual(currentView.bindingStartIndex, null, 'bindingStartIndex');
```
Comment (vicb) [inline packages/core/src/render3/instructions.ts]: Please change `'bindingStartIndex'` for something more helpful (i.e. `'the binding system should be initialized'`)

File: packages/core/src/render3/object_literal.ts (hunk 12 above)
Comment (vicb) [reply packages/core/src/render3/object_literal.ts]: :+1:

File: packages/core/test/render3/compiler_canonical_spec.ts (hunk 13 above)
Comment (vicb) [reply packages/core/test/render3/compiler_canonical_spec.ts]: When the template is `{'quoted': true, unquoted: false}` the (un)quoting must be preserved in the factory function.

File: packages/core/src/render3/instructions.ts (hunk 16 above)
Comment (kara) [reply packages/core/src/render3/instructions.ts]: The arg here is actually passed into a larger message built by `assertThrow`. So the constructed message would be:

```
ASSERT: expected bindingStartIndex !== null but was null
```

File: packages/core/test/render3/compiler_canonical_spec.ts (hunk 13 above)
Comment (kara) [reply packages/core/test/render3/compiler_canonical_spec.ts]: I see. I'll change one of the object literal keys to have quotes to ensure the spec is clear for the compiler.

File: packages/core/src/render3/object_literal.ts [hunk 17]
```diff
@@ -6,268 +6,285 @@
...
+ * @param factoryFn A factory function that takes binding values and builds an object or array
+ * containing those values.
+ * @param exp An array of binding values
+ * @returns A copy of the object/array or NO_CHANGE
+ */
+export function objectLiteralV(factoryFn: (v: any[]) => any, exps: any[]): any {
+  let different = false;
+
+  for (let i = 0; i < exps.length; i++) {
+    let exp = exps[i];
```
Comment (vicb) [inline packages/core/src/render3/object_literal.ts]: const

File: packages/core/src/render3/object_literal.ts (hunk 12 above)
Comment (kara) [reply packages/core/src/render3/object_literal.ts]: re perf, `peekBinding` should prime the cache, so `bind` will be a cache hit. Negligible performance impact.

**Review by vicb — APPROVED** [review]

**Comment by kara** [comment]
[presubmit](https://test.corp.google.com/ui#id=OCL:185467439:BASE:185467476:1518488020100:165ea256)
